### PR TITLE
[XGBoost] Add CUDA 12 Support, bump xgboost to v1.7.5

### PR DESCRIPTION
**Pull request by tylerjthomas9**
Ref: https://github.com/JuliaPackaging/Yggdrasil/issues/6174#issuecomment-1546072718

Changes in this PR:
- Add CUDA v12.0 build
- Add CUDA Arches to build (this is not needed, but I noticed it [here](https://github.com/JuliaPackaging/Yggdrasil/pull/6452), and it seems to be a [good practice](https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/)).
- Bump xgboost from v1.7.4 to v1.7.5
- Add CUDA support for linux + aarch64

### CODE REVIEW

**Comment by tylerjthomas9** [comment]
I think that this PR is ready to go

CC @maleadt  in case you're interested in reviewing the CUDA 12 addition

File: X/XGBoost/build_tarballs.jl [hunk 1]
```diff
@@ -57,38 +77,29 @@ fi
...
-# Dependencies that must be installed before this package can be built
 dependencies = [
     # For OpenMP we use libomp from `LLVMOpenMP_jll` where we use LLVM as compiler (BSD
     # systems), and libgomp from `CompilerSupportLibraries_jll` everywhere else.
     Dependency(PackageSpec(name="CompilerSupportLibraries_jll", uuid="e66e0078-7015-5450-92f7-15fbd957f2ae"); platforms=filter(!Sys.isbsd, platforms)),
     Dependency(PackageSpec(name="LLVMOpenMP_jll", uuid="1d63c593-3942-5779-bab2-d838dc0a180e"); platforms=filter(Sys.isbsd, platforms)),
-
-    # You can only specify one cuda version in the deps. To build against more than 
-    # one cuda version, you have to include them as Archive Sources. (see Torch_jll)
-    BuildDependency(PackageSpec(name="CUDA_full_jll", version=cuda_full_versions[cuda_version]), platforms=cuda_platforms),
```
Comment (maleadt) [inline X/XGBoost/build_tarballs.jl:91]: How are you building against CUDA without the toolchain?

**Comment by ViralBShah** [comment]
Should XGBoost+CUDA be a separate package? Or does it use CUDA only if it is found?

File: X/XGBoost/build_tarballs.jl (hunk 1 above)
Comment (ViralBShah) [reply X/XGBoost/build_tarballs.jl:91]: @maleadt Can you point to an example on how to do this right?

Comment (tylerjthomas9) [reply X/XGBoost/build_tarballs.jl:91]: Whenever I try to build with more than one `CUDA_full_jll` version, everything runs locally, but then it fails on buildkite.

For example, here is my second commit's build: https://buildkite.com/julialang/yggdrasil/builds/3050#018830ce-3515-460c-947f-f84c6e10133d

All CUDA versions error with this: 
```
[21:48:02] CMake Error at /usr/share/cmake/Modules/CMakeDetermineCUDACompiler.cmake:179 (message):
--
  | [21:48:02]   Failed to find nvcc.
  | [21:48:02]
  | [21:48:02]   Compiler requires the CUDA toolkit.  Please set the CUDAToolkit_ROOT
  | [21:48:02]   variable.
  | [21:48:02] Call Stack (most recent call first):
  | [21:48:02]   CMakeLists.txt:148 (enable_language)
```

I may have just been using the dependencies wrong, so I will revert the changes that put the `CUDA_full_jll` as an ArchiveSource and see if it works.

**Comment by tylerjthomas9** [comment]
> Should XGBoost+CUDA be a separate package? Or does it use CUDA only if it is found?

The only CUDA version is on machines with CUDA 11 (11+12 after this PR) and x64 linux. All other versions will not use CUDA. However, other platforms with CUDA seem to [sometimes pull in `CUDA_runtime_jll`](https://github.com/dmlc/XGBoost.jl/issues/156#issuecomment-1415809244), so we may have to separate the packages to fix this.

**Comment by ViralBShah** [comment]
I have done packages where I just do pkg and pkg_GPU. I find that it keeps things simpler while the packaging system is evolving for GPUs and such.

**Comment by maleadt** [comment]
> other platforms with CUDA seem to [sometimes pull in `CUDA_runtime_jll`](https://github.com/dmlc/XGBoost.jl/issues/156#issuecomment-1415809244), so we may have to separate the packages to fix this.

That's a PackageCompiler-related issue; you can ask it not to pull in lazy artifacts.

Ideally there should be no reason to have a separate GPU package. If the user doesn't have a GPU, you get the `cuda=none` tag and CUDA_Runtime_jll doesn't pull-in a huge artifact. Only if there's significant API differences between both builds it's really needed to have separate packages.

**Comment by ViralBShah** [comment]
Thanks @maleadt - that is good to know.

File: X/XGBoost/build_tarballs.jl (hunk 1 above)
Comment (maleadt) [reply X/XGBoost/build_tarballs.jl:91]: Sorry, missed your comment here. CUDA_full_jll isn't being installed when building for CUDA platforms, e.g. try `julia build_tarballs.jl --verbose x86_64-linux-gnu-libgfortran5-cxx11-cuda+12.0` and see how `CUDA_full_jll` isn't listed as part of the packages that are being installed (or pass `--debug` and see how `nvcc` isn't available). I'm not entirely sure why, but you did structure your `build_tarballs.jl` differently then how other packages do it. See e.g. AMGX, Extrae, PAPI, NVTX, etc.

Comment (tylerjthomas9) [reply X/XGBoost/build_tarballs.jl:91]: > See e.g. AMGX, Extrae, PAPI, NVTX, etc.

Ah, thank you. I will restructure it to follow the this style.

File: X/XGBoost/build_tarballs.jl [hunk 2]
```diff
@@ -95,7 +95,7 @@ for cuda_version in versions_to_build, platform in platforms
     end
 
     augmented_platform = Platform(arch(platform), os(platform);
-        cuda=isnothing(cuda_version) ? "none" : CUDA.platform(cuda_version)
+        cuda=isnothing(cuda_version) ? nothing : CUDA.platform(cuda_version)
```
Comment (maleadt) [inline X/XGBoost/build_tarballs.jl]: Why this change? With `nothing`, *any* artifact is selected. I think you really want `"none"`.

**Comment by tylerjthomas9** [comment]
This PR should be good to go. Building on Windows with CUDA support works, so I can add that in a future PR. However, we are still having issues with running XGBoost on Windows machines, so I don't see the need to add it right now.

**Comment by maleadt** [comment]
> Building on Windows with CUDA support works

How? AFAIU, that can't work on Yggdrasil because nvcc isn't a cross-compiler and instead expects MSVC.

**Comment by tylerjthomas9** [comment]
> How? AFAIU, that can't work on Yggdrasil because nvcc isn't a cross-compiler and instead expects MSVC.

You are right, I had a typo in the logic to test if its a cuda build (not in this main script but my test one), and it was just building for cpu. I was surprised when it built without errors, so I should've investigated to see if something was wrong.

**Comment by tylerjthomas9** [comment]
When refactoring, I accidentally changed `CUDA_Runtime_jll` to a BuildDependency instead of RuntimeDependency.

File: X/XGBoost/build_tarballs.jl [hunk 3]
```diff
@@ -57,43 +59,61 @@ fi
...
+    augmented_platform = Platform(arch(platform), os(platform);
+        cuda=isnothing(cuda_version) ? "none" : CUDA.platform(cuda_version)
+    )
+    should_build_platform(triplet(augmented_platform)) || continue
+
+    dependencies = AbstractDependency[
+        # For OpenMP we use libomp from `LLVMOpenMP_jll` where we use LLVM as compiler (BSD
+        # systems), and libgomp from `CompilerSupportLibraries_jll` everywhere else.
+        Dependency(PackageSpec(name="CompilerSupportLibraries_jll", uuid="e66e0078-7015-5450-92f7-15fbd957f2ae"); platforms=filter(!Sys.isbsd, [augmented_platform])),
+        BuildDependency(PackageSpec(name="LLVMOpenMP_jll", uuid="1d63c593-3942-5779-bab2-d838dc0a180e"); platforms=filter(Sys.isbsd, [augmented_platform])),
```
Comment (giordano) [inline X/XGBoost/build_tarballs.jl:106]: Why this was changed to build dep? This broke macos

**Comment by giordano** [comment]
This is also not expanding the C++ string ABI anymore
